python/lsst/ip/diffim/imageDecorrelation.py

```diff
@@ -353,3 +353,95 @@ def _doConvolve(exposure, kernel):
+        """Perform decorrelation operation on `subExposure`, using
+        `expandedSubExposure` to allow for invalid edge pixels arising from
+        convolutions.
+
+        This method performs A&L decorrelation on `subExposure` using
+        local measures for image variances and PSF. `subExposure` is a
+        sub-exposure of the non-decorrelated A&L diffim. It also
+        requires the corresponding sub-exposures of the template
+        (`template`) and science (`science`) exposures.
+
+        Parameters
+        ----------
+        subExposure : afw.Exposure
+            the sub-exposure of the diffim
+        expandedSubExposure : afw.Exposure
+            the expanded sub-exposure upon which to operate
+        fullBBox : afwGeom.BoundingBox
+            the bounding box of the original exposure
+        template : afw.Exposure
+            the corresponding sub-exposure of the template exposure
+        science : afw.Exposure
+            the corresponding sub-exposure of the science exposure
+        alTaskResult : pipeBase.Struct
+            the result of A&L image differencing on `science` and
+            `template`, importantly containing the resulting
+            `psfMatchingKernel`. Can be `None`, only if
+            `psfMatchingKernel` is not `None`.
+        psfMatchingKernel : Alternative parameter for passing the
+            A&L `psfMatchingKernel` directly.
+        kwargs :
+            additional keyword arguments propagated from
+            `ImageMapReduceTask.run`.
+
+        Returns
+        -------
+        A `pipeBase.Struct containing the result of the `subExposure`
+        processing, labelled 'subExposure'. It also returns the
+        'decorrelationKernel', although that currently is not used.
+
+        Notes
+        -----
+        This `run` method accepts parameters identical to those of
+        `ImageMapperSubtask.run`, since it is called from the
+        `ImageMapperTask`.  See that class for more information.
+        """
+        logLevel = self.log.getLevel()
+        self.log.setLevel(lsst.log.ERROR)  # Prevent too much log verbosity from DecorrelateALKernelTask.run
+        templateExposure = kwargs.get('template', None)  # input template
+        scienceExposure = kwargs.get('science', None)    # input science image
+        alTaskResult = kwargs.get('alTaskResult', None)
+        psfMatchingKernel = kwargs.get('psfMatchingKernel', None)
+        preConvKernel = kwargs.get('preConvKernel', None)
+
+        # subExp and expandedSubExp are subimages of the (un-decorrelated) diffim!
+        # So here we compute corresponding subimages of im1, and im2
+        subExp2 = scienceExposure.Factory(scienceExposure, expandedSubExposure.getBBox())
+        subExp1 = templateExposure.Factory(templateExposure, expandedSubExposure.getBBox())
+
+        psfMatchingKernel = alTaskResult.psfMatchingKernel if alTaskResult is not None else psfMatchingKernel
```

Comment:
I debated about that, but I think ValueError is for a different situation (e.g, received `-1` when only positive numbers are allowed): 

https://docs.python.org/2/library/exceptions.html#exceptions.ValueError
https://docs.python.org/2/library/exceptions.html#exceptions.RuntimeError